src/com/facebook/buck/worker/WorkerProcess.java

```diff
@@ -221,90 +213,88 @@ public synchronized void close() {
-        this.protocol.receiveNextCommandResponse();
+    Preconditions.checkState(
+        !Thread.holdsLock(this),
+        "About to block on input, should not be holding the lock that prevents new jobs");
+    WorkerProcessProtocol.CommandResponse commandResponse = protocol.receiveNextCommandResponse();
     SettableFuture<Integer> result = commandExitCodes.remove(commandResponse.getCommandId());
     Preconditions.checkState(
         result != null,
-        "Received message id %d with no corresponding waiter! (result was %d)",
+        "Received message id %s with no corresponding waiter! (result was %s)",
         commandResponse.getCommandId(),
         commandResponse.getExitCode());
 
     result.set(commandResponse.getExitCode());
   }
 
   private void readerLoop() {
-    try {
-      while (readerShutdownSignal.getCount() > 0) {
-        // A dance here to avoid calling `processNextCommandResponse` if we're not waiting for any
-        // commands. In the best case scenario, where we don't call `close()` until all commands are
-        // done, this will avoid painful exceptions in `close()`.
-        synchronized (commandExitCodes) {
-          while (readerShutdownSignal.getCount() > 0 && commandExitCodes.isEmpty()) {
-            commandExitCodes.wait();
+    boolean readerAlive = true;
+    while (true) {
+      // A dance here to avoid calling `processNextCommandResponse` if we're not waiting for any
+      // commands. In the best case scenario, where we don't call `close()` until all commands are
+      // done, this will avoid painful exceptions in `close()`.
+      //
+      // This uses `readerThread` instead of `this` as the lock so `close()` can block out
+      // `submitJob`, but still wait for this thread to exit.
+      synchronized (readerThread) {
+        while (!shutdownReaderThread && commandExitCodes.isEmpty()) {
+          try {
+            readerThread.wait();
+          } catch (InterruptedException e) {
+            throw new RuntimeException(e);
           }
         }
-        if (readerShutdownSignal.getCount() > 0 && !commandExitCodes.isEmpty()) {
-          processNextCommandResponse();
+        if (shutdownReaderThread) {
+          break;
         }
       }
-    } catch (Throwable t) {
-      // The WorkerProcessProtocol isn't really conducive to concurrency, so we just assume
-      // that shutdowns will cause some kind of exception.
-      // TODO(mikekap): Refactor the protocol enough to allow interrupting reads on a shutdown
-      // signal.
-      if (readerShutdownSignal.getCount() == 0) {
-        return;
-      }
-      LOG.error(t, "Worker pool reader process failed");
-    }
-  }
 
-  private void watchdogLoop() {
-    try {
-      while (!watchdogShutdownSignal.await(5, TimeUnit.SECONDS)) {
-        if (!readerThread.isAlive()) {
-          failAllFutures();
-        }
+      if (!readerAlive) {
+        failAllFutures();
+        continue;
```

Comment:
Once readerAlive becomes false, it stays false forever. Should it be set to true here?

Reply:
I'm confused with the readerAlive logic. Could you explain?

Reply:
Yep that's the intent. `readerAlive` is set to `false` only if the worker returns a malformed response. In this context, malformed responses are signs something terrible happened - even build step failures are well formed responses. This could only happen on e.g. malformed json or if the process just died. In that case, we stop attempting to communicate with the process at all and just fail all incoming requests until the build eventually `close()`s the process.

Reply:
I c, thank you.